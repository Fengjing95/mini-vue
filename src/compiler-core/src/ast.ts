/*
 * @Date: 2022-08-15 18:34:26
 * @Author: 枫
 * @LastEditors: 枫
 * @description: description
 * @LastEditTime: 2022-08-15 21:49:23
 */
export const enum NodeTypes {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT
}
