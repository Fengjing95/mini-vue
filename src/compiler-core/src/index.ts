/*
 * @Date: 2022-08-20 20:55:58
 * @Author: 枫
 * @LastEditors: 枫
 * @description: 统一出口
 * @LastEditTime: 2022-08-20 21:45:34
 */

export * from './compile'
