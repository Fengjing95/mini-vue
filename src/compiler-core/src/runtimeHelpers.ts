/*
 * @Date: 2022-08-18 20:23:40
 * @Author: 枫
 * @LastEditors: 枫
 * @description: helpers 常量
 * @LastEditTime: 2022-08-18 20:27:50
 */
export const TO_DISPLAY_STRING = Symbol('toDisplayString')

export const helperMapName: Record<symbol, string> = {
  [TO_DISPLAY_STRING]: 'toDisplayString'
}
