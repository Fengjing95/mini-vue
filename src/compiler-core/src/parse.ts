/*
 * @Date: 2022-08-15 18:05:45
 * @Author: 枫
 * @LastEditors: 枫
 * @description: parse
 * @LastEditTime: 2022-08-18 20:16:27
 */

import { NodeTypes } from './ast'

type ContextType = { source: string }

const enum TagType {
  TAG_START,
  TAG_END
}

export function baseParse(content: string) {
  const context = createParseContext(content)

  return createRoot(parseChildren(context, []))
}

function parseChildren(context: ContextType, ancestors: any[]) {
  const nodes = []

  while (!isEnd(context, ancestors)) {
    let node
    const s = context.source
    if (s.startsWith('{{')) {
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors)
      }
    }

    if (!node) {
      // 如果不是插值和 element
      node = parseText(context)
    }

    nodes.push(node)
  }
  return nodes
}

function isEnd(context: ContextType, ancestors: any[]) {
  // 遇到结束标签
  const s = context.source
  if (ancestors && s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag
      if (startsWithEndTagOpen(s, tag)) {
        return true
      }
    }
  }

  // source 没有值
  return !s
}

function parseText(context: ContextType) {
  let endTokens = ['<', '{{']
  let endIndex = context.source.length

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i])
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }

  const content = parseTextData(context, endIndex)
  return {
    type: NodeTypes.TEXT,
    content
  }
}

function parseTextData(context: ContextType, length: number) {
  // 获取 content
  const content = context.source.slice(0, length)

  // 推进
  advanceBy(context, content.length)
  return content
}

function parseElement(context: ContextType, ancestors: any[]) {
  // 解析 tag
  const element: any = parseTag(context, TagType.TAG_START)
  // 标签栈
  ancestors.push(element)
  element.children = parseChildren(context, ancestors)
  ancestors.pop()

  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagType.TAG_END)
  } else {
    throw new Error(`缺少闭合标签:${element.tag}`)
  }

  return element
}

function startsWithEndTagOpen(source: string, tag: string) {
  // 1. 头部 是不是以  </ 开头的
  // 2. 看看是不是和 tag 一样
  return (
    startsWith(source, '</') &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  )
}

function startsWith(source: string, searchString: string): boolean {
  return source.startsWith(searchString)
}

function parseTag(context: ContextType, type: TagType) {
  // 匹配标签
  // /<(\w+)[^>]*>(.*?<\/\1>)?/
  const match = /^<(\/?[a-z]*)/i.exec(context.source)
  const tag = match[1]

  // 删除解析完的代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)

  if (type === TagType.TAG_END) {
    return
  }
  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

function parseInterpolation(context: ContextType) {
  const openDelimiter = '{{'
  const closeDelimiter = '}}'

  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  ) // 闭合标识的索引
  advanceBy(context, openDelimiter.length) // 去除起始表示

  const rawContextLength = closeIndex - openDelimiter.length // 变量名长度
  // const rawContent = context.source.slice(0, rawContextLength) // 取出变量名
  const rawContent = parseTextData(context, rawContextLength)
  const content = rawContent.trim()

  advanceBy(context, closeDelimiter.length) // 去掉整个插值表达式

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content
    }
  }
}

// 推进
function advanceBy(context: ContextType, length: number) {
  context.source = context.source.slice(length)
}

function createRoot(children: any) {
  return { children, type: NodeTypes.ROOT }
}

function createParseContext(content: string): ContextType {
  return {
    source: content
  }
}
