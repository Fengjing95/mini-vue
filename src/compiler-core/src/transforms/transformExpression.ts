/*
 * @Date: 2022-08-18 19:38:07
 * @Author: 枫
 * @LastEditors: 枫
 * @description: 处理 ast 的插值表达式 simple 类型
 * @LastEditTime: 2022-08-18 20:22:23
 */
import { NodeTypes } from '../ast'

export function transformExpression(node: any) {
  if (node.type === NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content)
  }
}

function processExpression(node: any): any {
  node.content = `_ctx.${node.content}`
  return node
}
