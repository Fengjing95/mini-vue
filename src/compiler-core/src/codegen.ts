/*
 * @Date: 2022-08-17 22:27:19
 * @Author: 枫
 * @LastEditors: 枫
 * @description: 代码生成
 * @LastEditTime: 2022-08-18 20:34:28
 */

import { NodeTypes } from './ast'
import { helperMapName, TO_DISPLAY_STRING } from './runtimeHelpers'

export function generate(ast: any) {
  const context = createCodegenContext()
  const { push } = context

  genFunctionPreamble(ast, context)

  push('export ')

  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}) {`)

  push(`return `)
  genNode(context, ast.codegenNode)
  push('}')

  return { code: context.code }
}

function genFunctionPreamble(ast: any, context: any) {
  const VueBinging = 'vue'
  const aliasHelpers = (s: symbol) =>
    `${helperMapName[s]} as _${helperMapName[s]}`

  context.push(
    ast.helpers.length > 0
      ? `import { ${ast.helpers
          .map(aliasHelpers)
          .join(', ')} } from "${VueBinging}"\n`
      : ''
  )
}

function genNode(context: any, node: any) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(context, node)
      break

    case NodeTypes.INTERPOLATION:
      genInterpolation(context, node)
      break

    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(context, node)
      break

    default:
      break
  }
}

function genExpression(context: any, node: any) {
  const { push } = context
  push(`${node.content}`)
}

function genInterpolation(context: any, node: any) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(context, node.content)
  push(')')
}

function genText(context: any, node: any) {
  const { push } = context
  push(`'${node.content}'`)
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source: string) {
      context.code += source
    },
    helper(key: symbol) {
      return `_${helperMapName[key]}`
    }
  }

  return context
}
