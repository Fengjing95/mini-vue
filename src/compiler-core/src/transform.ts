import { NodeTypes } from './ast'
import { TO_DISPLAY_STRING } from './runtimeHelpers'

/*
 * @Date: 2022-08-17 17:34:50
 * @Author: 枫
 * @LastEditors: 枫
 * @description: 处理 ast 树
 * @LastEditTime: 2022-08-18 20:29:10
 */
export function transform(root: any, options: any = {}) {
  const context = createTransformContext(root, options)
  // 修改
  // 遍历--深度优先
  traverseNode(root, context)

  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0]
}

function traverseNode(node: any, context: any) {
  const nodeTransforms = context.nodeTransforms
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i]
    transform(node)
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context)
      break

    default:
      break
  }
}

function traverseChildren(node: any, context: any) {
  const children = node.children

  if (children) {
    // 递归遍历
    for (let i = 0; i < children.length; i++) {
      const node = children[i]
      traverseNode(node, context)
    }
  }
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key: string) {
      context.helpers.set(key, 1)
    }
  }

  return context
}
