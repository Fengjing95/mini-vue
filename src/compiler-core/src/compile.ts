/*
 * @Date: 2022-08-20 18:48:52
 * @Author: 枫
 * @LastEditors: 枫
 * @description: 模板编译
 * @LastEditTime: 2022-08-20 20:55:26
 */

import { generate } from './codegen'
import { baseParse } from './parse'
import { transform } from './transform'
import { transformElement } from './transforms/transformElement'
import { transformExpression } from './transforms/transformExpression'
import { transformText } from './transforms/transformText'

export function baseCompile(template: string) {
  const ast: any = baseParse(template)
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText]
  })

  return generate(ast)
}
