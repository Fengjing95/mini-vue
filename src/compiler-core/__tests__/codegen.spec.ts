import { generate } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { transformExpression } from '../src/transforms/transformExpression'

/*
 * @Date: 2022-08-17 22:24:22
 * @Author: 枫
 * @LastEditors: 枫
 * @description: description
 * @LastEditTime: 2022-08-18 20:12:34
 */
describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi')
    transform(ast)
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })

  it('interpolation', () => {
    const ast = baseParse('{{message}}')
    transform(ast, {
      nodeTransforms: [transformExpression]
    })
    const { code } = generate(ast)
    expect(code).toMatchSnapshot()
  })
})
