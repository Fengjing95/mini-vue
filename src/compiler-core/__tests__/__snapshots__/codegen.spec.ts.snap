// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen interpolation 1`] = `
"import { toDisplayString as _toDisplayString } from \\"vue\\"
export function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"export function render(_ctx, _cache) {return 'hi'}"`;
